### PR TITLE
[jit] Add support for __getstate__/__setstate__ on module

### LINKED ISSUE #19030: Support __get_state__/__set_state__ for JIT serialization
## 🚀 Feature
For custom C classes or for complex modules in JIT, there needs to be ability to supply customer serialization code.

## Motivation

We want to support custom opaque representations with the ability to serialize them. This is pretty important for opaque optimized layouts in inference (e.g. when using fbgemm), currently we use a hack with pack/unpack: https://github.com/pytorch/pytorch/blob/master/torch/jit/quantized.py#L27

This functionality will also be useful for custom C class bindings which would need to be serialized as well.

## Pitch

Proposal is to implement get_state/set_state similarly how they are handled in pickle: one can return any value (`IValue`) that will be serialized and later passed to `__set_state__` for deserialization.

```python
class MyLayerPacked(torch.nn.Module):
    # same as before
    def __init__(self, original: MyLayer):
        self.packed_weight = self.__pack__(original.weight) # ByteTensor with pointer
    def forward(self, x):
        return torch.my_mm_fast(self.packed_weight, x)
    # it has to be script method to work in C++
    @torch.jit.script_method
    def __getstate__(self):
        return {'weight': torch.unpack_from_my_mm(self.packed_weight)}
    def __setstate__(self, state):
        # attribute mutability is required here
        self.packed_weight = torch.pack_for_my_mm(state['weight'])
```

## Alternatives

Instead of following Pickle style we can define some other interface for it

### PR DESCRIPTION
**Pull request by driazati**
Adds support for `__getstate__` and `__setstate__` on modules that are called as part of export (`torch.save()`) and import (`torch.jit.load`).
* `__getstate__` and `__setstate__` must be TorchScript functions with the signatures `() -> T` and `(T) -> None` respectively
* The results of `__getstate__` are stored using the pickler in `states.pkl` with one for each module in definition order (`__getstate__` returns `None` by default if an imlpementation is not provided)
    * This prevents sharing between `__getstate__` and attributes, but this should be fine since their use is mostly unrelated (attributes are for storing values to be used in script methods, `__getstate__` for running arbitrary computations during import)

Follow up
* Somehow replacing `__getstate__`/`__setstate__` with a `ScriptMethodStub` makes `MyScriptModule().__getstate__()` call `ScriptModule.__getstate__()` when used in Python. This should be fixed so semantics in Python are preserved, but it doesn't affect the typical usage.

Fixes #19030

Differential Revision: [D15287161](https://our.internmc.facebook.com/intern/diff/15287161/)

### CODE REVIEW

File: torch/csrc/jit/export.cpp [hunk 1]
```diff
@@ -536,6 +538,9 @@ class ScriptModuleSerializer final {
 
   std::vector<IValue> attribute_table_;
 
+  // result of module __getstate__
```
Comment (zdevito) [inline torch/csrc/jit/export.cpp]: I think there should only be one table and one pkl file. Otherwise cross references between objects will get broken. Can you add a test for this as well?

File: torch/csrc/jit/export.cpp [hunk 2]
```diff
@@ -788,6 +796,9 @@ void ScriptModuleSerializer::convertModule(
     record->set_key(filename.str());
   }
 
+  // Run the '__getstate__' method on the module and store the result
```
Comment (zdevito) [inline torch/csrc/jit/export.cpp]: The behavior here seems weirdly organized. getstate should error if it doesn't exist and we should be checking here if it exists. Furthermore, if module has a getstate/setstate then we _should not_ serialize its attributes and tensors, I do not see code for that. It is likely that having getstate corresponds with having tensors and attributes that _are not serializable_ so this distinction is important.

File: torch/csrc/jit/import.cpp [hunk 3]
```diff
@@ -303,6 +309,13 @@ void ScriptModuleDeserializer::convertModule(
         tensor_table_,
         import_callback);
   }
+
+  // TODO: get the correct index so this works for submodules
```
Comment (zdevito) [inline torch/csrc/jit/import.cpp]: this TODO should be in the patch.

File: torch/csrc/jit/import.cpp [hunk 4]
```diff
@@ -303,6 +309,13 @@ void ScriptModuleDeserializer::convertModule(
...
+  static size_t module_num = 0;
+  if (module_num < state_table_.size()) {
+    module->setstate(state_table_.at(module_num));
+  }
+  module_num++;
```
Comment (zdevito) [inline torch/csrc/jit/import.cpp]: Similar to the reasons for attributes containing an ID, this should not simply update a number. It is hard to follow and couples the serialization format with the deserialization order. What if we wanted to only deserialize one submodule?

File: torch/csrc/jit/script/module.h [hunk 5]
```diff
@@ -328,6 +328,92 @@ struct TORCH_API Module {
 
     return nullptr;
   }
+
+  void validateGetSetStateSchema() const {
```
Comment (zdevito) [inline torch/csrc/jit/script/module.h]: I am not a fan of how long this error checking code is, but I don't have a better solution for you.

File: torch/csrc/jit/script/module.h [hunk 6]
```diff
@@ -328,6 +328,92 @@ struct TORCH_API Module {
...
+          get_type->python_str(),
+          " does not match '__setstate__'s argument type (",
+          set_type->python_str(),
+          "))");
+    }
+  }
+
+  IValue getstate() const {
+    if (find_method("__getstate__") == nullptr) {
+      return IValue();
```
Comment (zdevito) [inline torch/csrc/jit/script/module.h]: Should throw if not present similar to python.

File: torch/csrc/jit/script/module.h [hunk 7]
```diff
@@ -328,6 +328,92 @@ struct TORCH_API Module {
...
+
+    validateGetSetStateSchema();
+
+    Stack stack;
+    stack.emplace_back(module_object());
+    getstate->run(stack);
+    return stack.at(0);
+  }
+
+  void setstate(IValue state) {
```
Comment (zdevito) [inline torch/csrc/jit/script/module.h]: I don't think these should be members of the module.h class. We do not treat members specially, even forward doesn't have a function. It makes more sense to put this logic in the serializer.

**Review by zdevito — COMMENTED** [review]
Looks almost ready to go. A bunch of little comments.

File: torch/csrc/jit/script/module.h (hunk 5 above)
Comment (driazati) [reply torch/csrc/jit/script/module.h]: It'd be nice if we could specify a target schema string to match to, but that's outside the scope of this PR.

File: torch/csrc/jit/export.cpp (hunk 2 above)
Comment (driazati) [reply torch/csrc/jit/export.cpp]: If we only serialize one or the other, how to we handle references to parameters/attributes? Should we be serializing only the proto stubs with the names/types and then leaving their values uninitialized and assume that `__setstate__` will handle them? Providing a `__getstate__` may also only mean that some of the attributes/tensors are not serializable, but maybe some of the others are. It seems potentially confusing if some features are being turned off by adding a `__getstate__` method.

File: torch/csrc/jit/script/python_sugared_value.cpp [hunk 8]
```diff
@@ -330,6 +330,35 @@ std::vector<std::shared_ptr<SugaredValue>> ModuleValue::asTuple(
   return result;
 }
 
+void ModuleValue::setAttr(
```
Comment (zdevito) [inline torch/csrc/jit/script/python_sugared_value.cpp]: This code is essentially a reimplementation of setAttr on the SimpleValue class when the type is a class.  But `self_` is already a class. We should change ModuleValue so that it inherits from SimpleValue. Then it would also inherit the correct definition of setAttr and no code would be needed here.

File: caffe2/proto/torch.proto [hunk 9]
```diff
@@ -69,6 +69,9 @@ message ModuleDef {
   optional bool optimize = 8;
 
   repeated AttributeDef attributes = 9;
+
+  // Used for retrieving module state from the pickled IValues table
+  optional int64 id = 10;
```
Comment (zdevito) [inline caffe2/proto/torch.proto]: The name is more generic than its purpose, maybe `get_state_attribute_id`?

Comment (zdevito) [reply caffe2/proto/torch.proto]: Did we bump the appropriate version number when adding this field?

File: torch/csrc/jit/export.cpp [hunk 10]
```diff
@@ -745,22 +830,39 @@ void ScriptModuleSerializer::convertModule(
...
+  }
   for (const auto& attribute : module.get_attributes()) {
     // Add attribute to ModuleDef
     torch::AttributeDef* attribute_def = module_def->add_attributes();
     attribute_def->set_name(attribute.name());
     attribute_def->set_type(attribute.type()->python_str());
 
-    attribute_table_.push_back(attribute.value());
-    attribute_def->set_id(attribute_table_.size() - 1);
+    if (!user_provided_serialization) {
```
Comment (zdevito) [inline torch/csrc/jit/export.cpp]: We should not write out parameters as well. This features exists because there is a non-serializable tensor format being used for a parameter.

File: torch/csrc/jit/import.cpp [hunk 11]
```diff
@@ -255,6 +259,24 @@ void ScriptModuleDeserializer::importCallback(const std::string& qualifier) {
...
+  auto setstate = module->class_compilation_unit().find_function("__setstate__");
+
+  AT_CHECK(
+      setstate != nullptr,
+      "Cannot call '__setstate__' method because"
+      " it does not exist");
+
+  // TODO: once modules are first class in the interpreter and methods are not
+  // lowered, change this to `module->run_method("__setstate__", state);`
+  Stack stack;
```
Comment (zdevito) [inline torch/csrc/jit/import.cpp]: `Stack stack = { module->module_object(),  state}; `

File: torch/csrc/jit/script/sugared_value.h [hunk 12]
```diff
@@ -114,11 +114,12 @@ struct TORCH_API SimpleValue : public SugaredValue {
     else
       return NEVER;
   }
-  std::vector<std::shared_ptr<SugaredValue>> asTuple(
+  virtual std::vector<std::shared_ptr<SugaredValue>> asTuple(
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.h]: virtual is redundant here. I think some compilers complain about adding it.

File: torch/csrc/jit/script/sugared_value.h [hunk 13]
```diff
@@ -141,6 +142,10 @@ struct TORCH_API SimpleValue : public SugaredValue {
     return value_;
   }
 
+  virtual std::shared_ptr<SimpleValue> new_with(Value* new_value) {
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.h]: in person: let's just make ModuleValue::setAttr do `SimpleValue(self_).setAttr` that would avoid these issues of needing to propagate SimpleValue subclasses around.

**Review by zdevito — APPROVED** [review]
Looks good. Two minor things: parameters need to not be saved as well as attributes, and SimpleValue changes.

**Comment by dzhulgakov** [comment]
Thanks, @driazati ! Are you planning to follow up and replace existing __pack__ / __unpack__ usages? (or @jamesr66a)
